### PR TITLE
Fix for bug when using Basehook.get_uri() with a postgres connection

### LINKED ISSUE #27474: Use full postgresql URI connection prefix.
### Apache Airflow version

2.4.2

### What happened

This creates a deprecated URI string. I've added a new connection, using `get_uri()` returns a URI with `postgres://` instead of `postgressql://` so it's incompatible with a few libs. Should this be updated to return valid format URI ?

### What you think should happen instead

The URI should return URI with `postgresql://` prefix.

### How to reproduce

Add a new postgres connection, pull up the Connection by ID in DAG and try using it directly with SQLAlchemy

### Operating System

Debian

### Versions of Apache Airflow Providers

_No response_

### Deployment

Docker-Compose

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by andersbogsnes**
When using Basehook.get_uri() with a postgres connection, the scheme of the returned URI will be postgres instead of postgresql. SQLAlchemy expects postgres URIs to have the `postgresql` and errors when trying to use URIs generated in this way

I've implemented the bugfix with the simplest possible change, though I imagine the "proper" fix would be to rename the Postgres connection type to `postgresql` though I imagine this would require more extensive rework as well as potentially breaking existing installations

closes: #27474

### CODE REVIEW

File: airflow/models/connection.py [hunk 1]
```diff
@@ -212,7 +212,9 @@ def get_uri(self) -> str:
                 self.conn_type,
             )
 
-        uri = f"{str(self.conn_type).lower().replace('_', '-')}://"
+        # SQLAlchemy expects postgresql:// as scheme
+        scheme = "postgresql" if self.conn_type == "postgres" else self.conn_type
+        uri = f"{str(scheme).lower().replace('_', '-')}://"
```
Comment (uranusjr) [inline airflow/models/connection.py]:
```suggestion
        self.conn_type == "postgres":
            scheme = "postgresql"
        else:
            scheme = self.conn_type.lower().replace("_", "-")
        uri = f"{scheme}://"
```

No need to do the lower-replace combo on the “postgresql” literal.

Comment (andersbogsnes) [reply airflow/models/connection.py]: Sure - was aiming for minimum extra logic, but this is also fine :+1:

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Itnermittent problems with integration tests in sqlite. Mergine.

**Comment by Taragolis** [comment]
@potiuk @uranusjr Just wondering is `airflow.models.connection.Connection.get_uri()` intend to return valid SQLAlchemy URI? Might be I wrong but I thought this method only use for generate Airflow Connection URI.

And we could be in a situation if user generate Airflow Connection URI for PostgreSQL and try to reuse it, e.g. as environment variable than common-sql-provider operators might not work because hook register with `postgres` conn_type not as `postgresql`

And as far as I know `PostgresHook` has own method
https://github.com/apache/airflow/blob/fb7c6afc8cb7f93909bd2e654ea185eb6abcc1ea/airflow/providers/postgres/hooks/postgres.py#L170-L178

**Comment by potiuk** [comment]
> And we could be in a situation if user generate Airflow Connection URI for PostgreSQL and try to reuse it, e.g. as environment variable than common-sql-provider operators might not work because hook register with `postgres` conn_type not as `postgresql`

@Taragolis  - I think you are right. I was hesitating a bit before approving that one (for that very reason) - but I could not figure out a case where this could go wrong, so I figured that is OK to get it in.  There were few people trying to do the same and reported similar issue, so I figured that we should be ok to do that (even though it was a bit strange to implement postgresl-specific thing in BaseHook.

And yeah, that changes my "approval" vote for that scenario. @andersbogsnes I think that would be interesting to see why in this case you cannot use PostgresHook's method - and why you expect to have valid SQLAlchemy URL here (which is different than Airflow URL)..

**Review by potiuk — CHANGES_REQUESTED** [review]
Doubts raised by @Taragolis need to be sorted out - I am now not sure if this is a good idea.

**Comment by Taragolis** [comment]
> but I could not figure out a case where this could go wrong, so I figured that is OK to get it in.

Just check and seems like nothing actually bad happen because we have replace from `postgresql` to `postgres` during parsing Airflow Connection URI to URI Object. So PostgreSQL connection is unique connection which have some fixes in core parts.

https://github.com/apache/airflow/blob/b18f3190c6c157301cb7d29c439dddbb9a8dbf14/airflow/models/connection.py#L182-L188


```python
import os
from airflow.models.connection import Connection
from airflow.providers.common.sql.operators.sql import BaseSQLOperator

PG_CONN_ID = "pg_conn"

conn = Connection(
    conn_id=PG_CONN_ID,
    conn_type="postgres",
    host="foo.bar",
    login="spam",
    password="egg",
)
print(f"AIRFLOW_CONN_{conn.conn_id.upper()}='{conn.get_uri()}'")
# Current Impl: AIRFLOW_CONN_PG_CONN='postgres://spam:egg@foo.bar'
# This PR Impl: AIRFLOW_CONN_PG_CONN='postgresql://spam:egg@foo.bar'


for ix, conn_url in enumerate(["postgres://spam:egg@foo.bar", "postgresql://spam:egg@foo.bar"], start=1):
    conn_parsed = Connection(conn_id="dummy", uri=conn_url)
    print(f"{ix}. Airflow Connection URI {conn_url!r}, parsed conn type {conn_parsed.conn_type!r}")
    # 1. Airflow Connection URI 'postgres://spam:egg@foo.bar', parsed conn type 'postgres'
    # 2. Airflow Connection URI 'postgresql://spam:egg@foo.bar', parsed conn type 'postgres'

    os.environ[f"AIRFLOW_CONN_{PG_CONN_ID.upper()}"] = conn_url
    op = BaseSQLOperator(task_id="sample", conn_id=PG_CONN_ID)
    hook = op.get_db_hook()
    print(f"{ix}. Hook: {type(hook)}, URI {hook.get_uri()}")
    # 1. Hook: <class 'airflow.providers.postgres.hooks.postgres.PostgresHook'>, URI postgresql://spam:egg@foo.bar
    # 2. Hook: <class 'airflow.providers.postgres.hooks.postgres.PostgresHook'>, URI postgresql://spam:egg@foo.bar
```


> There were few people trying to do the same and reported similar issue

I hear about `https` schema for HTTP connection and questions why Connection.get_uri() does not generate "https://foo.bar"

**Comment by andersbogsnes** [comment]
I don't have a strong opinion either way, it's more of a conceptual issue of what a URI means and what end users can expect to use that for. 

My team ran into it because they had an expectation that they can directly use an Airflow uri as a connection string, which is a general requirement if we want to support using f.ex a Postgres connection to inject the DB url into a KubernetesPodOperator via Jinja Templating.

I would argue that getting a uri should be consistent across public APIs or explicitly documented as being for one purpose or the other. 

Which is correct in this case, I'll leave up to you - it's not blocking anything since there's a simple workaround as pointed out previously. The core is the end-user "surprise" at the difference between the APIs

**Comment by potiuk** [comment]
I'd say we leave it. I really do not like the fact that we are adding "postgres" specific thing in "Connection". Connection should be unaware of the types of connections it can have.

There is a "proper" way of getting the postgresql. And the surprise is pretty much built-in in Airflow "differences" between different kinds of URLs.

**Pull request closed without merging**